calico/felix/fiptables.py

```diff
@@ -1,214 +1,451 @@
+        then the rule is appended.
+
+        If force_position is True, then the rule is added at the specified point
+        unless it already exists there. If force_position is False, then the
+        rule is added only if it does not exist anywhere in the list of rules.
+
+        Position starts at offset 0, even though iptables wants it to start at
+        offset 1.
+        """
+        if position == RULE_POSN_LAST:
+            position = len(self.rules)
+
+        if force_position:
+            if (len(self.rules) <= position) or (rule != self.rules[position]):
+                # Either adding after existing rules, or replacing an existing rule.
+                self._insert_rule(rule, position)
+        else:
+            if rule not in self.rules:
+                self._insert_rule(rule, position)
+        return
+
+    def _insert_rule(self, rule, position):
+        """
+        Insert the given rule at the supplied position.
+        
+        position of 0 means "insert at the start", which is what iptables
+        considers as position 1.
+
+        This method is private; callers outside this module should call
+        insert_rule instead.
+        """
+        args = [IPTABLES_CMD[self.type],
+                "-w",
+                "-t",
+                self.table.name,
+                "-I",
+                self.name,
+                str(position + 1)]
+        args.extend(rule.generate_fields())
+        self.table.ops.append(args)
+        self.rules.insert(position, rule)
+        log.debug("Creating rule : %s", args)
+
+    def __str__(self):
+        output = "  Chain %s\n" % self.name
+        for rule in self.rules:
+            output += "    %s\n" % rule
+
+        return output
+
+class Table(object):
+    """
+    Representation of an iptables table.
+    """
+    def __init__(self, type, name):
+        self.type = type
+        self.name = name
+        self.chains = {}
+        self.ops = []
+
+    def is_chain(self, name):
+        return (name in self.chains)
+
+    def get_chain(self, name):
+        """
+        Gets a chain, creating it first if it does not exist.
+        """
+        chain = self.chains.get(name)
+
+        if chain is None:
+            log.debug("Creating chain %s in table %s (%s)",
+                      name,
+                      self.name,
+                      self.type)
+            chain = Chain(self, name)
+            self.ops.append([IPTABLES_CMD[self.type],
+                              "-w",
+                              "-t",
+                              self.name,
+                              "-N",
+                              name])
+
+        return chain
+
+    def delete_chain(self, chain_name):
+        log.debug("Delete chain %s from table %s (%s)",
+                  chain_name,
+                  self.name,
+                  self.type)
+        self.ops.append([IPTABLES_CMD[self.type],
+                        "-w",
```

Comment:
You're right. I was being dim (I checked the man page, but clearly incompetently since it says what you say, not what I remembered).

I'll raise an issue and fix it up.